File: Daulis/Read.cs
using System;
using System.IO;
using System.Numerics;
using System.Text;

namespace Daulis
{
    public static class Read
    {
        public static int ReadParticles(Particle[] ps, string fileName)
        {
            Complex pos = 0;
            for (int j = 0; j < ps.Length; j++)
            {
                ps[j] = new Particle();
            }

            int i = 0;
            using var streamReader = new StreamReader(fileName, Encoding.ASCII);
            for (char chr = Convert.ToChar(streamReader.Read());
                !streamReader.EndOfStream;
                chr = Convert.ToChar(streamReader.Read()))
            {
                if (chr == '\n')
                {
                    // Since input particles are of height 2, newline needs
                    // to increment y by 2.
                    pos += 2;
                    pos = pos.Real;

                    continue;
                }

                if (chr > ' ')
                {
                    // We've detected an actual particle. We double up the
                    // height to 2.
                    ps[i].Position = pos;
                    ps[i + 1].Position = pos + 1;

                    // The # character signifies a 'wall' (solid barrier.)
                    ps[i].IsWall = ps[i + 1].IsWall = (chr == '#');

                    i += 2;
                }

                pos -= Complex.ImaginaryOne;
            }

            return i;
        }
    }
}

File: Daulis.FSharp/Read.fs
module Daulis.FSharp.Read

open System
open System.IO
open System.Numerics
open System.Text

let fromFile len (filename: string) =
    let mutable position: Complex =
        Complex.op_Implicit 0

    let particles =
        Array.create len Particle.Default

    let mutable length = 0

    use streamReader =
        new StreamReader(filename, Encoding.ASCII)

    while not streamReader.EndOfStream do
        let chr =
            streamReader.Read() |> Convert.ToChar

        match chr with
        | '\n' ->
            position <- position + 2.0
            position <- Complex.op_Implicit position.Real
        | _ when chr > ' ' ->
            particles[length] <- { particles[length] with Position = position }
            particles[length + 1] <- { particles[length] with Position = position + 1.0 }
            particles[length] <- { particles[length] with IsWall = chr = '#' }
            particles[length + 1] <- { particles[length + 1] with IsWall = chr = '#' }
            length <- length + 2
            position <- position - Complex.ImaginaryOne
        | _ -> position <- position - Complex.ImaginaryOne

    (particles, length)
